packages/react/src/utils/useIntersectionRef.ts

```diff
@@ -0,0 +1,66 @@
+import type { MutableRefObject } from 'react';
+import { useRef, useEffect } from 'react';
+
+/**
+ * When a component needs to track intersection via a ref, useIntersectionRef
+ * will return a ref object containing the results from IntersectionObserver
+ * for the current intersection entry.
+ *
+ * @example
+ * const elementRef = useRef<HTMLElement>()
+ * const intersectionRef = useIntersectionRef<HTMLElement>(elementRef)
+ * return <span ref={elementRef}>...</span>
+ */
+export default function useIntersectionRef<T extends HTMLElement>(
+  element: T | MutableRefObject<T>,
+  intersectionObserverOptions: IntersectionObserverInit = {
+    root: null,
+    threshold: 1.0
+  }
+): MutableRefObject<IntersectionObserverEntry | null> {
+  const intersectionRef = useRef<IntersectionObserverEntry | null>(null);
+
+  useEffect(() => {
+    // istanbul ignore else
+    if (
+      'IntersectionObserver' in globalThis &&
+      typeof IntersectionObserver === 'function'
+    ) {
+      if (
+        !element &&
+        typeof element === 'object' &&
```

Comment:
Trying to understand the logic here. How could `!element` be together with `typeof element === 'object'`?

Reply:
`null` values, although it could be simplified as the 3rd statement is redundant and we cover non HTMLElement values below.